### PR TITLE
Send user agent

### LINKED ISSUE #329: Set a webtau user agent by default and allow overriding in cfg

### PR DESCRIPTION
**Pull request by tsiq-karold**
Closes #329 

Default behaviour is to send `webtau/<version>`.  If a user agent is specified in cfg, we send `<cfg-user-agent> (webtau/<version>)`.  It is also possible to not append the `(webtau/<version>)` bit with a custom agent by specifying an appropriate cfg option.

### CODE REVIEW

**Comment by tsiq-rob** [comment]
> I thought it would be useful to include the webtau part and version even with a custom agent.

as a counter argument, this will probably annoy someone who specifically wants to override the user agent with something of their own choosing

**Comment by tsiq-karold** [comment]
Yeah, that's fair.  I considered allowing both but I don't want the cfg to get too crazy.

File: webtau-config/src/main/java/com/twosigma/webtau/cfg/WebTauConfig.java [hunk 1]
```diff
@@ -46,6 +46,8 @@
     private final ConfigValue waitTimeout = declare("waitTimeout", "wait timeout in milliseconds", () -> 5000);
     private final ConfigValue disableFollowingRedirects = declareBoolean("disableRedirects", "disable following of redirects from HTTP calls");
     private final ConfigValue maxRedirects = declare("maxRedirects", "Maximum number of redirects to follow for an HTTP call", () -> 20);
+    private final ConfigValue userAgent = declare("userAgent", "User agent to send on HTTP requests",
+            () -> "webtau/"+getClass().getPackage().getImplementationVersion());
```
Comment (tsiq-clemens) [inline webtau-config/src/main/java/com/twosigma/webtau/cfg/WebTauConfig.java]: a couple of spaces around the + make it prettier

File: webtau-config/src/main/java/com/twosigma/webtau/cfg/WebTauConfig.java [hunk 2]
```diff
@@ -180,6 +182,19 @@ public int maxRedirects() {
         return maxRedirects.getAsInt();
     }
 
+    public String getUserAgent() {
+        if (userAgent.isDefault()) {
+            return userAgent.getAsString();
+        }
+
+        String defaultValue = userAgent.getDefaultValue().toString();
+        return userAgent.getAsString() + "/" + defaultValue;
```
Comment (tsiq-clemens) [inline webtau-config/src/main/java/com/twosigma/webtau/cfg/WebTauConfig.java]: would it look better to have the webtau/version in (), e.g.   partner-api-tests (webtau/1.8).
vs partner-api-tests/webtau/1.8, or what if people have a slash in their agent config? api/v2/webtau/1.8

I might be overthinking this.

Comment (tsiq-karold) [reply webtau-config/src/main/java/com/twosigma/webtau/cfg/WebTauConfig.java]: I think you might be right, I prefer it in () than after a /.

**Comment by tsiq-karold** [comment]
Hopefully this latest version satisfies both @tsiq-clemens and @tsiq-rob 's requirements/comments 😄

File: pom.xml [hunk 3]
```diff
@@ -610,6 +610,19 @@
                     </dependency>
                 </dependencies>
             </plugin>
+
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-jar-plugin</artifactId>
```
Comment (MykolaGolubyev) [inline pom.xml]: what is this for?

File: webtau-config/src/main/java/com/twosigma/webtau/cfg/WebTauConfig.java [hunk 4]
```diff
@@ -46,6 +46,11 @@
     private final ConfigValue waitTimeout = declare("waitTimeout", "wait timeout in milliseconds", () -> 5000);
     private final ConfigValue disableFollowingRedirects = declareBoolean("disableRedirects", "disable following of redirects from HTTP calls");
     private final ConfigValue maxRedirects = declare("maxRedirects", "Maximum number of redirects to follow for an HTTP call", () -> 20);
+    private final ConfigValue userAgent = declare("userAgent", "User agent to send on HTTP requests",
+            () -> "webtau/" + getClass().getPackage().getImplementationVersion());
```
Comment (MykolaGolubyev) [inline webtau-config/src/main/java/com/twosigma/webtau/cfg/WebTauConfig.java]: I think we expose version in some other way

File: webtau-config/src/main/java/com/twosigma/webtau/cfg/WebTauConfig.java [hunk 5]
```diff
@@ -46,6 +46,11 @@
...
+    private final ConfigValue removeWebtauFromUserAgent = declare("removeWebtauFromUserAgent",
```
Comment (MykolaGolubyev) [inline webtau-config/src/main/java/com/twosigma/webtau/cfg/WebTauConfig.java]: was wondering why enable is a toggle and disable is default. It is a bit unusual for REST api to expect user agent.

File: webtau-http-groovy/src/test/groovy/com/twosigma/webtau/http/HttpGroovyTest.groovy [hunk 6]
```diff
@@ -239,6 +240,41 @@ class HttpGroovyTest implements HttpConfiguration {
...
+        } finally {
+            cfg.userAgentConfigValue.reset()
+        }
+    }
+
+    @Test
+    void "custom user agent without webtau and its version"() {
+        try {
+            cfg.userAgentConfigValue.set("test", "custom")
+            cfg.removeWebtauFromUserAgent.set("true", true)
```
Comment (MykolaGolubyev) [inline webtau-http-groovy/src/test/groovy/com/twosigma/webtau/http/HttpGroovyTest.groovy]: first parameter value is the source, not sure what "true" means in this case

**Comment by MykolaGolubyev** [comment]
A bit late to the party, but left a few comments.

File: pom.xml (hunk 3 above)
Comment (tsiq-karold) [reply pom.xml]: To have the version populated in the jar manifest and available via `getClass().getPackage().getImplementationVersion())`

File: webtau-config/src/main/java/com/twosigma/webtau/cfg/WebTauConfig.java (hunk 4 above)
Comment (tsiq-karold) [reply webtau-config/src/main/java/com/twosigma/webtau/cfg/WebTauConfig.java]: I didn't find it but if it is already available I can undo the jar packaging change and switch this line over.

File: webtau-config/src/main/java/com/twosigma/webtau/cfg/WebTauConfig.java (hunk 5 above)
Comment (tsiq-karold) [reply webtau-config/src/main/java/com/twosigma/webtau/cfg/WebTauConfig.java]: I agree that it's unlikely an API will expect an agent but it's useful for identifying requests from tests in logs.

File: webtau-http-groovy/src/test/groovy/com/twosigma/webtau/http/HttpGroovyTest.groovy (hunk 6 above)
Comment (tsiq-karold) [reply webtau-http-groovy/src/test/groovy/com/twosigma/webtau/http/HttpGroovyTest.groovy]: Oops, meant to say `test` like the line above.

**Pull request merged**